### PR TITLE
Remove active_support Time extensions

### PR DESCRIPTION
**Pull request by dduugg**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----
As requested in https://github.com/Homebrew/brew/issues/10508, I was looking at the performance of requiring various `ActiveSupport` `require` statements, and noticed that 35-39% of the time in `global.rb` was spent in `require "active_support/core_ext/numeric/time"` (as measured via `brew prof --stackprof help`):

image

This PR removes that require statement, and cleans up related code.

The long pole of `ActiveSupport` `require` statements is now `require "active_support/core_ext/string/inflections"`, at 6.9-7.8% of time in `global.rb`:

image

### CODE REVIEW

**Review by MikeMcQuaid — APPROVED** [review]
Good catch, thanks again @dduugg!

Also: check your email, I sent you a message 😁

**Comment by MikeMcQuaid** [comment]
> The long pole of `ActiveSupport` `require` statements is now `require "active_support/core_ext/string/inflections"`, at 6.9-7.8% of time in `global.rb`:

Would be game to remove that next, I think!

**Comment by dduugg** [comment]
Note that this is a risky change. I found some use of the extensions that weren't covered by tests by `grep`ping for the most common 🐵 patches, but could easily have missed others.

Note that I also tried removing the parts of the `ActiveSupport` `rbi` file that were concerned with time/duration to see if that would trigger any type errors (for demonstration only, reverted in a subsequent commit in this PR).

File: Library/Homebrew/global.rb [hunk 1]
```diff
@@ -13,7 +13,7 @@
 # Only require "core_ext" here to ensure we're only requiring the minimum of
 # what we need.
 require "active_support/core_ext/object/blank"
-require "active_support/core_ext/numeric/time"
+require "active_support/core_ext/string/filters"
```
Comment (dduugg) [inline Library/Homebrew/global.rb:16]: This was a transitive `require` that we rely on in a few places (in particular, `String#remove`).

File: Library/Homebrew/test/cleanup_spec.rb [hunk 2]
```diff
@@ -37,8 +37,8 @@
     end
 
     it "returns true when ctime and mtime < days_default" do
-      allow_any_instance_of(Pathname).to receive(:ctime).and_return(2.days.ago)
-      allow_any_instance_of(Pathname).to receive(:mtime).and_return(2.days.ago)
+      allow_any_instance_of(Pathname).to receive(:ctime).and_return((DateTime.now - 2).to_time)
```
Comment (dduugg) [inline Library/Homebrew/test/cleanup_spec.rb:40]: I prefer this style of subtracting days from a `Time` object (it's less math-y), but there's a different, existing style elsewhere in this file (`Time.now - (2 * 60 * 60 * 24)`). I'm happy to use that one for consistency, if reviewers prefer.

Comment (issyl0) [reply Library/Homebrew/test/cleanup_spec.rb:40]: I prefer your way, but then I'm allergic to maths. 😉

Comment (apainintheneck) [reply Library/Homebrew/test/cleanup_spec.rb:40]: I don't think it matters in this case but I don't think those two options are 100% similar.

```rb
irb(main):009:0> (Time.now - (2 * 60 * 60 * 24))
=> 2023-02-02 18:17:39 -0800
irb(main):010:0> (DateTime.now - 2).to_time
=> 2023-02-02 18:18:00 -0800
```

I prefer the `DateTime` version too.

**Review by apainintheneck — APPROVED** [review]
LGTM! I took a quick look and couldn't find any remaining references to these extensions. I was also able to see a small speedup when benchmarking it on my old iMac.

#### Before
```
/u/l/Homebrew (master|✔) $ hyperfine 'brew ls'
Benchmark 1: brew ls
  Time (mean ± σ):      2.175 s ±  0.044 s    [User: 1.005 s, System: 0.961 s]
  Range (min … max):    2.111 s …  2.250 s    10 runs
```

#### After
```
/u/l/Homebrew (active-support-reduce|✔) [1]$ hyperfine 'brew ls'
Benchmark 1: brew ls
  Time (mean ± σ):      1.841 s ±  0.061 s    [User: 0.841 s, System: 0.814 s]
  Range (min … max):    1.783 s …  2.000 s    10 runs
```

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
Thanks again @dduugg!

**Comment by p-linnane** [comment]
This caused breakage when referencing `time.rfc3339`:
```
Error: An exception occurred within a child process:
  NoMethodError: undefined method `rfc3339' for 2023-02-02 08:18:45 UTC:Time
```
Reverting this locally resolved the issue.

**Comment by dduugg** [comment]
@p-linnane thanks for flagging, and apologies for the breakage. I've proposed a fix in https://github.com/Homebrew/homebrew-core/pull/122480 (and also listed some other approaches to resolve).

**Comment by chenrui333** [comment]
I wonder if we can just revert this PR and retry with a better fix for https://github.com/Homebrew/homebrew-core/pull/122480 (mostly for the sugar syntax stuff)

**Pull request merged**